lib/progressive_hash_diff.rb

```diff
@@ -0,0 +1,29 @@
+class ProgressiveHashDiff
+  attr_reader :presenter, :previous_item, :current_item
+
+  def initialize(presenter)
+    @presenter = presenter
+  end
+
+  def previous_item=(previous_item)
+    @previous_item = presenter ? presenter.call(previous_item) : previous_item
+  end
+
+  def current_item=(current_item)
+    @current_item = presenter ? presenter.call(current_item) : current_item
+  end
+
+  def diff
+    raise MissingItemError, 'No previous item provided' if @previous_item.nil?
+    raise MissingItemError, 'No current item provided' unless @current_item.present?
```

Comment:
I've switched l18 to use `if item.nil?` -- `present?` won't work for l17 as is needs to be at least an empty hash, and definitely not nil.